archunit/src/test/java/com/tngtech/archunit/library/modules/ArchModulesTest.java

```diff
@@ -0,0 +1,511 @@
+                        from("module1", "sub1").to("module1", "sub2")
+                                .withClassDependencies(from(FirstClassInSubModule11.class).to(FirstClassInSubModule12.class)
+                                        .from(FirstClassInSubModule11.class).to(SecondClassInSubModule12.class)),
+                        from("module1", "sub1").to("module2", "sub1")
+                                .withClassDependencies(from(SecondClassInSubModule11.class).to(FirstClassInSubModule21.class)));
+    }
+
+    @Test
+    public void provides_module_dependencies_to_self() {
+        ArchModules<?> modules = ArchModules
+                .defineByPackages(validTestExamplePackage + ".(*).(*)")
+                .modularize(testExamples);
+
+        ArchModule<?> module = modules.getByIdentifier("module2", "sub1");
+
+        assertThatModuleDependencies(module.getModuleDependenciesToSelf())
+                .containOnlyModuleDependencies(
+                        from("module1", "sub1").to("module2", "sub1")
+                                .withClassDependencies(from(SecondClassInSubModule11.class).to(FirstClassInSubModule21.class)),
+                        from("module1", "sub2").to("module2", "sub1")
+                                .withClassDependencies(from(FirstClassInSubModule12.class).to(FirstClassInSubModule21.class)));
+    }
+
+    @Test
+    public void all_dependencies_not_covered_by_module_dependencies_are_considered_undefined() {
+        ArchModules<?> modules = ArchModules
+                .defineByPackages(validTestExamplePackage + ".(*).(*)")
+                .modularize(testExamples);
+
+        ArchModule<?> module = modules.getByIdentifier("module1", "sub1");
+
+        assertThatDependencies(module.getUndefinedDependencies())
+                .contain(FirstClassInSubModule11.class, String.class)
+                .contain(FirstClassInSubModule11.class, List.class)
+                .contain(SecondClassInSubModule11.class, Collection.class)
+                .contain(SecondClassInSubModule11.class, FirstClassInModule2.class)
+                .doesNotContain(FirstClassInSubModule11.class, FirstClassInSubModule12.class)
+                .doesNotContain(FirstClassInSubModule11.class, SecondClassInSubModule12.class);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> T getValue(Field staticField) {
+        try {
+            return (T) staticField.get(null);
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private String getExamplePackage(String subpackageName) {
+        return getClass().getPackage().getName() + ".testexamples." + subpackageName;
+    }
+
+    private static ModuleDependenciesAssertion assertThatModuleDependencies(Collection<? extends ModuleDependency<?>> dependencies) {
+        return new ModuleDependenciesAssertion(dependencies);
+    }
+
+    static class ModuleDependenciesAssertion extends
```

Comment:
I guess you mean for discoverability in the future? But I'm not completely convinced. Because often I feel we ponder a lot about the production code to be modular and encapsulated, then in the test scope we create some monolithic entangled mess where one place knows all the contexts again :thinking: That ship might already have sailed with `testutil.assertion`, but I still have my doubts that this specific library stuff should end up there :thinking: Likely a convention where modularized assertions somehow mimic the production structure might be better :thinking: I think as long as `testutil.assertion` only tackles core domain objects it would be okay. That we've mingled assertions for core and lang is already a little questionable, but also mingling in super specific library assertions ... I don't know :man_shrugging:

Reply:
Okay, convinced! 😉